src/sentry/migrations/0686_alert_rule_project_backfill_migration.py

```diff
@@ -0,0 +1,69 @@
+# Generated by Django 5.0.3 on 2024-03-22 22:36
+
+import logging
+
+from django.db import migrations
+
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapper
+
+logger = logging.getLogger(__name__)
+
+
+def _backfill_alert_rule_projects(apps, schema_editor):
+    QuerySubscriptions = apps.get_model("sentry", "QuerySubscription")
+    AlertRuleProjects = apps.get_model("sentry", "AlertRuleProjects")
+
+    # use RangeQuerySetWrapper to avoid loading all subscriptions into memory
+    for subscription in RangeQuerySetWrapper(QuerySubscriptions.objects.all()):
+
+        snuba_query = subscription.snuba_query
+        if not snuba_query:
+            logger.warning(
+                "QuerySubscription found with no snuba_query",
+                extra={"query_subscription_id": subscription.id},
+            )
+            continue
+
+        alert_rule = snuba_query.alertrule_set.get()
+
+        existing_alert_rule_projects = list(AlertRuleProjects.objects.filter(alert_rule=alert_rule))
+        should_create_new = not existing_alert_rule_projects.exists()
+
+        if not should_create_new and (  # eg. if alert rule already has fk projects
+            len(existing_alert_rule_projects) > 1
+            or existing_alert_rule_projects[0].project != subscription.project
```

Comment:
`existing_alert_rule_projects[0].project_id` save you a query.

Shouldn't we be checking all rows to see if the project doesn't match?

Reply:
the AlertRuleProject doesn't have a `project_id` attr though?
or do you mean i should be comparing the `project_id`s as opposed to the project?

I guess that's one way to do it
we could iterate through all the projects if the len(projects) >=1
Then delete any projects that don't match the subscription

now that we've listified the query, i suppose that makes sense 👍 
will update